hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrngSystemPrecompiledContract.java

```diff
@@ -0,0 +1,283 @@
+			final GasCalculator gasCalculator,
+			final PrngLogic prngLogic,
+			final ExpiringCreations creator,
+			final SideEffectsTracker tracker,
+			final RecordsHistorian recordsHistorian,
+			final PrecompilePricingUtils pricingUtils,
+			final Supplier<Instant> consensusNow,
+			final LivePricesSource livePricesSource,
+			final GlobalDynamicProperties properties) {
+		super(PRECOMPILE_NAME, gasCalculator);
+		this.prngLogic = prngLogic;
+		this.creator = creator;
+		this.tracker = tracker;
+		this.recordsHistorian = recordsHistorian;
+		this.pricingUtils = pricingUtils;
+		this.consensusNow = consensusNow;
+		this.livePricesSource = livePricesSource;
+		this.properties = properties;
+	}
+
+	@Override
+	public long gasRequirement(final Bytes bytes) {
+		return gasRequirement;
+	}
+
+	@Override
+	public PrecompileContractResult computePrecompile(final Bytes input, final MessageFrame frame) {
+		ExpirableTxnRecord.Builder childRecord;
+		Bytes result;
+		Bytes randomNum = Bytes.EMPTY;
+
+		try {
+			gasRequirement = calculateGas();
+			validateTrue(frame.getRemainingGas() >= gasRequirement, INSUFFICIENT_GAS);
+
+			final var selector = input.getInt(0);
+			randomNum = generatePseudoRandomData(selector, input);
+
+			childRecord = createSuccessfulChildRecord(randomNum, frame, input);
+			result = SUCCESS_RESULT;
+
+		} catch (InvalidTransactionException e) {
+			childRecord = createUnsuccessfulChildRecord(e.getResponseCode(), frame,
+					e.isReverting() ? e.getRevertReason() : Bytes.EMPTY);
+			result = resultFrom(e.getResponseCode());
+		} catch (Exception e) {
+			log.warn("Internal precompile failure", e);
+			childRecord = createUnsuccessfulChildRecord(FAIL_INVALID, frame, Bytes.EMPTY);
+			result = resultFrom(FAIL_INVALID);
+		}
+
+		final var parentUpdater = updater.parentUpdater();
+		if (parentUpdater.isPresent()) {
+			final var parent = (AbstractLedgerWorldUpdater) parentUpdater.get();
+			parent.manageInProgressRecord(recordsHistorian, childRecord, body(randomNum, input));
+		} else {
+			throw new InvalidTransactionException("PRNG precompile frame had no parent updater", FAIL_INVALID);
+		}
+
+		return result == null ?
+				PrecompiledContract.PrecompileContractResult.halt(null, Optional.of(ExceptionalHaltReason.NONE))
+				: PrecompiledContract.PrecompileContractResult.success(result);
+	}
+
+	Bytes generatePseudoRandomData(final int selector, final Bytes input) {
+		return switch (selector) {
+			case PSEUDORANDOM_SEED_GENERATOR_SELECTOR -> random256BitGenerator();
+			case PSEUDORANDOM_NUM_IN_RANGE_GENERATOR_SELECTOR -> randomNumberGeneratorInRange(input);
+			default -> null;
+		};
+	}
+
+	private Bytes random256BitGenerator() {
+		final var hashBytes = prngLogic.getNMinus3RunningHashBytes();
+		if (isEmptyOrNull(hashBytes)) {
+			return null;
+		}
+		return Bytes.wrap(hashBytes, 0, 32);
+	}
+
+	private Bytes randomNumberGeneratorInRange(final Bytes input) {
+		final var range = rangeValueFrom(input);
```

Comment:
⚠️ This will throw an `IllegalArgumentException` if `input` represents e.g. a `long` (much less `uint256`); and then we'll get `FAIL_INVALID` instead of `INVALID_PRNG_RANGE`.

Reply:
I think we are treating it as `uint256`, it solves a host for problems for mod.